### ISSUE TITLE
Simplify language used in documentation

### ISSUE THREAD

**Issue by ridiculousfish** [fish-shell/fish-shell#740]
This issue tracks reducing the complexity of fish's documentation according to a readability checker.

Bryan Kilgallin reports on the fish-users list:

Take the following quote as an example. I parsed it through a 
readability checker.
http://www.online-utility.org/english/readability_test_and_improve.jsp
That gave a Gunning Fog Index of 14, and a Flesch Reading Ease of 38. 
The following URL explains that this excerpt looks like a highbrow 
newspaper. Authors should try to make it more easily understood.

Variable expansion is the only type of expansion performed on double 
quoted strings. There is, however, an important difference in how 
variables are expanded when quoted and when unquoted. An unquoted 
variable expansion will result in a variable number of arguments. For 
example, if the variable $foo has zero elements or is undefined, the 
argument $foo will expand to zero elements. If the variable $foo is an 
array of five elements, the argument $foo will expand to five elements. 
When quoted, like "$foo", a variable expansion will always result in 
exactly one argument. Undefined variables will expand to the empty 
string, and array variables will be concatenated using the space character.

http://ridiculousfish.com/shell/user_doc/html/index.html#expand

**Comment by faho** [comment]
We should just hire [Randall Munroe](https://xkcd.com/1133/) for this.

**Comment by pickfire** [comment]
@faho, haha. He isn't simplifying the language, he is creating jokes and drawing stickmans.

**Comment by zachyordy** [comment]
Link in OP (http://ridiculousfish.com/shell/user_doc/html/index.html#expand) gives a 404. Has the link moved?

**Comment by faho** [comment]
@zachyordy: The fish documentation has moved to http://fishshell.com/docs/current/index.html from @ridiculousfish's personal site since his fork became the official one.

**Comment by faho** [comment]
One thing I'd like to know:

I kinda think the current style tries too hard to avoid addressing "you", and in doing so uses some constructs that I consider much too formal or "stuffy".

E.g.

>Every program on your computer can be used as a command in `fish`. If the program file is located in one of the directories in the <a href="#variables-special">`PATH`</a>, it is sufficient to type the name of the program to use it.

"It is sufficient" sounds to me like something [Teal'c](https://en.wikipedia.org/wiki/Teal'c) would say. Which is to say quite alien.

Now, I'm not a native speaker, but wouldn't it be nicer as

>If the program file is located in one of the directories in the <a href="#variables-special">`PATH`</a>, you can just type the name of the program to use it.

?

**Comment by ridiculousfish** [comment]
Yes it would, I agree

**Comment by sabinem** [comment]
I would like to take this over:
- for me this is a good way to get into fish
- as a newcomer, I can notice, what I do understand and what is not explained well enough.

My first impression regarding the documentation of fish is that the [tutorial](https://fishshell.com/docs/current/tutorial.html) is in better shape then the main [documentation](https://fishshell.com/docs/current/index.html):

What I miss the most in the main documentation is examples. I feel examples ease learning a lot. So they should be an important part in any documentation. The tutorial has gotten this right, but the documentation is just missing out in this regard.

My plan on improving the [documentation](https://fishshell.com/docs/current/index.html) would be:
- put in examples
- try to simplify the language and adapt it in style to the  [tutorial](https://fishshell.com/docs/current/tutorial.html)

What do you think about this?

**Comment by faho** [comment]
>What do you think about this?

@sabinem: Sure, go for it!

However, I wouldn't _overdo_ the examples. It's more important that what is there is understandable and correct than simply having _more_. Keep in mind that the main documentation isn't just for first-time readers unfamiliar with fish, but also useful as a reference for later.

**Comment by sabinem** [comment]
> It's more important that what is there is understandable and correct than simply having more. Keep in mind that the main documentation isn't just for first-time readers unfamiliar with fish, but also useful as a reference for later.

@faho: I understand. Can we at least agree, that I add only simple and tested examples, where I absolutely need them to simplify the explanation? 

For example:
- there is not a single example on escaping characters: it would be helpful to see at least one such example
- there is no example on redirecting I/O, there should be at least one such example
etc.

I promise I will be careful and test all examples I put in and I will go without examples if they are not needed for the explanation. 

Would that be acceptable for you?

(All documentations I know come with examples. I come from Python mostly, which has a lot of examples in its docs.)

**Comment by faho** [comment]
>Would that be acceptable for you?

Yes, very much so!

**Comment by sabinem** [comment]
I am working on this, but I struggle a bit in what is expected from me in this regard: 

I notice that I am not very happy with the main documentation as it is: in my humble opinion, it would need both:
- simplifying the language 
- restructuring of the sections

The restructuring of the sections is especially needed since the documentation does not provide search, so only the sections can be found easily.

## This post includes:
- an analysis of its current structure
- a proposal of its restructuring

I very much hope we can discuss this here, before I start working on this issue. 

## Analysis of the current structure

- **Introduction**
- **Syntax overview**
  - Quotes
  - Escaping characters
  - Input/Output (IO) redirection
  - Piping
  - Background jobs
  - Job control
  - Functions
    - Defining aliases
    - Autoloading functions
    - Conditional execution of code and flow control
  - Some common words
- **Autosuggestion**
- **Tab completion**
  - Writing your own completions
  - Useful functions for writing completions
  - Where to put completions
- **Parameter expansion (Globbing)**
  - Wildcards
  - Command substitution
  - Brace expansion
  - Variable expansion
  - Cartesian Products
  - Index range expansion
  - Home directory expansion
  - Combining different expansions
- **Shell variable and function names**
  - Variable scope
  - More on universal variables
  - Variable scope for functions
  - Exporting variables
  - Special variables
  - The status variable
  - Variables for changing highlighting colors
  - Locale variables
- **Shell variables**
- **Building commands**
- **Command line editor**
  - Shared bindings
  - Emacs mode commands
  - Vi mode commands
    - Command mode
    - Insert mode
    - Visual mode
  - Copy and paste (Kill Ring)
  - Searchable history
  - Multiline editing
- **Running multiple programs**
- **Initialization files**
- **Future feature flags**
- **Other features**
  - Syntax highlighting
  - Programmable title
  - Programmable prompt
  - Configurable greeting 
  - Event handlers
  - Debugging fish scripts
- **Further help and development**

**Comment by faho** [comment]
@sabinem:

>The restructuring of the sections is especially needed

I'm not opposed to restructuring (where it makes sense, of course), but I still believe that the more pressing concern is simplifying the language.

>since the documentation does not provide search

Your browser does, though. That has worked for me, in practice.

>many important topics are hidden away in subsections and are hard to find

Agreed!

>the structure is not balanced

Is that actually a concern? If there is not much to say about one topic, should it then be lumped together with other things or expanded?

I'm not really a writer, so I'm not sure if that is a common thing to strive for.

>for example piping is not only a matter of syntax

Yeeeaaahhh... nothing is "only" a matter of syntax, so I'm not sure what you're going for here.

>to me this seems like a organically grown documentation, that could stand some pruning and reordering

Most definitely!

>Introduction: this section is very short. Instead I would introduce here the POSIX standard and say what features are met by the fish shell and where it differentiates.

Please don't introduce POSIX sh here. For one, we imagine that this document would also be read by people not familiar with it, so that would throw them off the deep end. Also it would bloat this chapter _quite_ a bit. Listing fish's differences from POSIX is gonna take a bit.

I'd rather we explain what a shell is a bit more here.

>all the points discussed here are features of the shell rather than syntax.

Like I said, there is no "just" syntax. Every piece of syntax is going to have attached semantics, which means it's also a feature.

>I would call this Feature overview instead

I would be okay with that. We might want to distinguish "language features" (which is what I imagine with "syntax") from the others (like interactive things - keybindings etc) though.

>I would drop the subsections 'Quotes' and 'Escaping characters' and combine them into a section 'Strings' instead.

Okay, though both the terms "Quotes" and "Escaping" _must_ be used - that's what people would search for.

>Why are they not prefixed '__fish' if they are intended for external usage?

Like you said, organically grown documentation... We have an issue somewhere about renaming those functions.

>the status deserves to have its own section

Yup!

>I would call that Configurations instead

I'm not a native speaker, but "Configurations" sounds awkward to me. Just singular "Configuration"?

> I tried this command status featuresand got status: Subcommand 'features' is not valid: should that section be removed? I do not understand this.

It's new in 3.0, not in a release yet. Try building fish from git master and you should have it.

**Comment by kopischke** [comment]
I agree with a lot of @sabinem’s points: I was a pretty advanced `bash` user before switching to fish, and I found the documentation hard to follow due to its, um, quirky structure. To this day, I am often not sure in what section of the docs I need to look things up; browser search helps, but when search is your best option, structure is on indeterminate leave. And that is from someone accustomed to reading the friggin’ `bash` man page…

**Comment by sabinem** [comment]
@faho, @pickfire: Thank you both for your comments. It encourages me to work on this. Can you add the label `docs`to this issue and what about assigning this to me?

**Comment by pickfire** [comment]
@sabinem Thanks a lot for commenting this out. (I rarely help out with fish lately)

The proposal for restructuring looks good. But there is one section that I have been struggling so far (never really searched) where sharing configuration between computers for fish isn't simple, maybe some note about it is better when we use `abbr -a` or `set -Ux` which is one-time in fish. So far I have something like (thanks to @faho):
```fish
if not set -q fish_initialized
        abbr -a fish_uninitialized set -e fish_user_abbreviations\; set -e fish_initialized

        abbr -a sx      exec sx
        abbr -a gdb     gdb -q

        abbr -a pacman  sudo pacman

        set -U TZ       Asia/Singapore
        set -U CFLAGS   '-O2 -march=native -pipe'
        set -U CXXFLAGS $CFLAGS
        set -U MAKEFLAGS -j5

        set -U GOPATH   $HOME/src/go
        set -U GOROOT   /usr/lib/go
        set -U fish_user_paths ~/.local/bin $PATH
        set -U TIMEWARRIORDB ~/.task/timew

        set -U fish_initialized
end
```

As well, it would be nice to include some reference stuff (linking rare features) where couldn't be easily found in fish without looking into the source code for functions like say *hidden* keyboard shorcuts like <kbd>Alt</kbd> + <kbd>l</kbd>. We could mention a bit maybe in something like See Also or put some links to show them in the documentation.

**Comment by sabinem** [comment]
@pickfire: I don't really understand your post: 

> So far I have something like (thanks to @faho)

- what is that code for?

> say hidden keyboard shorcuts like Alt + 

- what does that keyboard shortcut do?

Do you have some links for for me, so I can reading up on the context?

**Comment by pickfire** [comment]
@sabinem Sorry for not being clear there.

> what is that code for?

That code is run only once where if fish is not initialized (the first time you ran) it will run that code. See https://github.com/faho/config/blob/master/fish/.config/fish/conf.d/abbrs.fish

>  what does that keyboard shortcut do?

Same as `ls --color`.

> Do you have some links for for me, so I can reading up on the context?

Links, nope and that is the reason why I mentioned it here.

**Comment by sabinem** [comment]
@pickfire : Thank you now I understand, what you mean, okay, I will try to add this to the documentation.

**Comment by sabinem** [comment]
I am working on the documentation: there are some parts that I do not understand, so I will ask here:

#### [Tab Completions](https://fishshell.com/docs/current/index.html#completion)

>  - Completion of filenames, even on strings with wildcards such as '*', '**' and '?'.

I tried completion with wildcard, but for me it did not work. I entered `ls user_doc/**` and hit the tab key: nothing happened. Can someone please tell me what I am getting wrong here.

#### Redirection with Filedescriptors N
- descripted in [Syntax Overview](https://fishshell.com/docs/current/index.html#syntax)

Any file descriptor can be redirected in an arbitrary way by prefixing the redirection with the file descriptor.

> To redirect input of FD N, write N<DESTINATION
> To redirect output of FD N, write N>DESTINATION
> To append the output of FD N to a file, write N>>DESTINATION_FILE

In fish FD > 2 does not work. Is that correct? 

In bash can open another file descriptor and attach it to a file like this:
```
> exec 3<> /tmp/foo    #open fd 3.
> echo "test" >&3
> exec 3>&-                  #close fd 3.
> cat /tmp/foo 
test
```
This does not work in fish. So am I correct that there are only 0,1,2 as FDs in fish? Or can someone explain to me how to use an FD 3 in fish?

Sorry to bother you with those questions, but in order to simplify the language in the documentation, I feel I need to understand the topics that are addressed there.

**Comment by faho** [comment]
>This does not work in fish. So am I correct that there are only 0,1,2 as FDs in fish? Or can someone explain to me how to use an FD 3 in fish?

Only 1 and 2 work _correctly_. Others don't behave sensibly, and the `exec >&X` thing is not supported - see #3303 and #3948.

>Can someone provide an example, where this is used?

Escaped characters? All the time!

Mostly, this is to prevent characters that are special to fish from being interpreted as special. E.g. to pass a literal `;` for use with `find`'s `-exec` option, you need `find . -something something -exec {} \;`. If you don't escape it, fish will use it as the command separator, so `find` won't get it.

Other sequences are used to generate characters that are otherwise annoying to include in your code, for example because they are non-printable.

E.g. our bindings use something like

```fish
bind $argv \n execute
```

to bind "newline" to the "execute" function - that (or `\r` for carriage return) is what you get when you press the return key.

Our completions also frequently use `\t` for the tab character, because that's the separator between completion candidate and description.

`\e` and `\c` are used all over the place for bindings (e.g. `\cs` for control-s), and `\e` also for things dealing with escape sequences. 

Much of this is used with `printf` or `string` instead of outside of quotes, but it's the same principle. All of these characters would be annoying to include - because they either have absolutely no visual representation (escape), or they just look like other whitespace (tab) or they break the flow (newline) or even obscure other characters (carriage return), and often awkward to type (we bind the tab key, so inserting a literal tab character needs some other way).

>echo \011 for example it just gives me \011

Urgh, that's unfortunate. It's `\0XY`, with a leading _zero_, not the letter Oh-as-in-Oscar.

```fish
test \011 = \t; and echo Both tabs!
```

I don't think I've seen one of these used, anywhere. If anything, people use `\X` and `\U`.

**Comment by sabinem** [comment]
@faho: Thanks a lot for taking so much time explaining this to me. That really helps!

**Comment by sabinem** [comment]
I have a question to the debugging section:

I established a breakpoint in a function: 
```fish
function hal2
    ls
    echo "hallo"
    echo "I am happy"
    breakpoint
end
```

I got a new prompt:  `BP hal2:3 >`, which is nice. I understand that `BP` means  Breakpoint and `hal2` was the name of my function, but what is `3`. Initially I sought it might be the linenumber of the breakpoint. But it is always 3, even if I add another breakpoint or move the breakpoint to another line. So why `3`? Hope someone can help me out on this. 

Also is there a way to skip breakpoint and exit immediately. So far I noticed `exit` just takes me from breakpoint to breakpoint.

**Comment by faho** [comment]
>Initially I sought it might be the linenumber of the breakpoint. But it is always 3, even if I add another breakpoint or move the breakpoint to another line. So why 3?

This is defined in the `fish_breakpoint_prompt` function. Unfortunately, the line number appears to be where the `status line-number` call is.

So it's a bug.

>How can I do this? Can you give me an example for this?

Theoretically, you'd `kill -TRAP PID_OF_THAT_FISH`. However, in my testing I have not been able to get that to work. Both running `sleep` (which of course is an external command) and `read` (which is a builtin), nothing happened when I sent the signal.

So I'm assuming you'd have to send the signal when the parser or something else from fish's core is running, which does not leave much room and makes this basically useless.

**Comment by sabinem** [comment]
@faho: So for now I will remove that part about setting breakpoints by trapping running jobs from the documentation. 

Is it okay to issue an "issue" about the prompt: I could not find one about that bug. 

Also by working on the documentation I am also trying things out and I notice  a lot of issues in the process. I feel there are parts of `fish` that work very well and there are other parts that are still quite bumpy. 

I hope you won't get annoyed by too many issues, but when I notice things that I feel are worth a discussion. I will post them as separate issues, okay?

**Comment by faho** [comment]
>I feel there are parts of fish that work very well and there are other parts that are still quite bumpy.

That's a fair summary, yes.

>I will post them as separate issues, okay?

Sure!

>Is it okay to issue an "issue" about the prompt: I could not find one about that bug.

I took a look at the source, and #4161 is already about `status line-number` not being useful, so I don't think we need another issue about this.

**Comment by sabinem** [comment]
I have a question regarding the documentation of process expansion: this first paragraph is just for the context.

> Process expansion
The % (percent) character at the beginning of a parameter followed by a string is expanded into a process ID (PID). The following expansions are performed:
If the string is the entire word self, the shell's PID is the result.
Otherwise, if the string is the entire word last, the last job's PID is the result.
Otherwise, if the string is the ID of a job, the result is the process group ID of the job.
Otherwise, if any child processes match the specified string, their PIDs are the result of the expansion.
Otherwise, if any processes owned by the user match the specified string, their PIDs are the result of the expansion.
If none of these matches apply, an error is produced.
This form of expansion is useful for commands like kill and fg, which take process IDs as arguments.

**Question:**
I could not make sense of these two options:

> Otherwise, if any child processes match the specified string, their PIDs are the result of the expansion.
Otherwise, if any processes owned by the user match the specified string, their PIDs are the result of the expansion.

Can you please provide an example for this, so that I understand better, what the match is meant to be in these two cases.

**Comment by ridiculousfish** [comment]
@sabinem This is rad! Thanks for taking this on!

The docs on `fishshell.com` refer to the most recently released version of fish. In master, process and job expansion is (for the moment) removed, and the docs in master have been updated.

What you can do is build the docs locally, `make doc` or `ninja doc`, however you build. You'll see that process and job expansion is no longer mentioned.

**Comment by sabinem** [comment]
@ridiculousfish : I had to look up what "rad" means. Thanks for the encouragement. 

You are right: process and job expansion are removed. 

This rewrite of the documentation is a lot of work and may take a while to finish. I hope the fish team will then be happy with the result once I am done. 

I picked that work, since I want to do a project in documentation and this one seems to me as a good match for what I was looking for.

I am just wondering: when will fish 3.0 get out? Do you have any date in mind for this? And might it make sense combining this with [#5063](https://github.com/fish-shell/fish-shell/issues/5063 )

Have you started that transition to Sphinx already? May I join into this? Just for fun I imported the documentation into Sphinx on my local computer and writing good texts it is a lot easier there, since I am better able to instantly see what I am doing in Sphinx. 

I am also wondering: Restructuring the documentation and rewriting the text in a simpler language, changes the documentation a great deal. Would it be better to combine that switch with a milestone? It may also take someone to approve the changed documentation once it is ready, since git will not be of much help showing changes almost everywhere.

Any thoughts on this?

**Comment by faho** [comment]
>I am just wondering: when will fish 3.0 get out? Do you have any date in mind for this?

In the near future. I wouldn't add any new feature to it.

>This rewrite of the documentation is a lot of work and may take a while to finish. 

If I may give some guidance: I wouldn't retreat into my cave for six months to "rewrite all the documentation". Instead, I'd recommend you do incremental, piece-by-piece changes to it as much as you can. That way we're able to discuss things before you've spent months going in the wrong direction. Which would then mean we'd reject your change and all involved would feel terrible.

So how about you pick a chapter or something, work on that and then present it?

**Comment by sabinem** [comment]
@faho : 

> If I may give some guidance: I wouldn't retreat into my cave for six months to "rewrite all the documentation". Instead, I'd recommend you do incremental, piece-by-piece changes to it as much as you can. That way we're able to discuss things before you've spent months going in the wrong direction. Which would then mean we'd reject your change and all involved would feel terrible.
So how about you pick a chapter or something, work on that and then present it?

That is indeed a very good idea and this will solve my problem. I will do it this way. Thank you very much for this suggestion!

**Comment by sabinem** [comment]
Just some suggestions on how I will proceed with this:

- I have already rewritten some bigger parts, for example job control, has been just 5 lines, but now it is a sections  with subsections 

- if I add that into the documentation right now, it will look as if this topic gets out of proportion. 

- therefore I will continue to work on my rewrite separately, to preserve the big picture for myself

- in parallel I will start with pull requests, that make tiny changes: one step at a time, gradually aiming to get to the point where my rewrite is right now and later

- so the PRs do not reflect how far I have gotten, but they will aim to make the process for both sides manageable

- I make one PR per structure change and then extra PRs, when I am changing content

- I will start this today with some structural changes.

Any further thoughts or suggestions on this?

**Comment by zanchey** [comment]
Sorry for not speaking up sooner. The documentation is [close to my heart](https://github.com/fish-shell/fish-shell/pull/702/files) and it's fantastic that you're interested in looking at it.

There is an awful lot of complexity contained within the manual and it is easy to lose some of the subtleties in a rewrite. You will get lots of review comments but I'm very grateful that you're having a go as it has been on my list for several years.

I agree that some restructuring is required. Your suggested outline looks good. The original poster (Bryan) emailed me extensively, and one of his suggestions was that fish's manual needs to be aimed at beginners. At least some explanation of what a shell is, the basic concepts of programs, builtins and functions, piping and redirection, and so on. Of course, you don't have to add this straight away! 

As @faho has said the documentation has very much grown organically. I think the [FreeBSD Documentation Project guidelines](https://www.freebsd.org/doc/en/books/fdp-primer/writing-style.html) are as good as any, although many of the more technical aspects involve their toolchain which is very different to ours.

Thanks for taking this on!

**Comment by sabinem** [comment]
@zanchey : your comment makes me happy: this is exactly how I feel about it: fish is a shell, that is very friendly to beginners and that should get reflected in the documentation. 

I already wrote that chapter on what a shall is and will add it to my first PR after what you said about the target group. Also my chapter about job control, assumes no prior knowledge, this is why it has subsections. I might add that too then. 

Maybe I will try to keep my first PR focused on a rewrite of these two first sections "Introductions" and "Syntax Overview". This will make for a start, where you can see the how I write this and I can get initial feedback on these very important first parts of the documentation.

**Comment by zanchey** [comment]
Sounds like a great plan!

**Comment by sabinem** [comment]
Would it be okay if I make a first PR with the following changes:

- just putting in **little content tables for all the subsections** that are hidden away: I will put these content tables in the sections directly under the section header and they will link to subsection headers

- make the few **subsubsections** into subsections

This would have the advantage that the content is overall more reachable and not hidden away. 

After this I will start with the content changes. I just don't want to do this together with content changes. And the content changes will still keep me busy for a while.

Tell me if this has any disadvantage, otherwise I will go ahead and prepare this PR.

**Comment by sabinem** [comment]
Just to let you know I am still working on this. It is hard for me to just come up with a sample.

I am trying to round things off, so that the whole documentation stays rounded and coherent and by that I am changing more and more sections. :(

I know it will be a risk and won't be offended if my pull request will get rejected on the first trial. Then it was just an exercise and will learn from it doing better on my next attempt to bundle the changes into smaller PRs, but at least then I will have a somewhat complete picture of the documentation in my mind: I will have a plan on where I want to take it.

Hoping for everybody to be patient with me. :) And as I say: I won't be disappointed if my changes get rejected initially, I probably need that draft for myself to better understand the topic at hand.

**Comment by sabinem** [comment]
I am more then half way through the documentation now. It is taking shape. I also notice that some parts are not described at all. Other things I can't get to work on my machine as they are described.

So I will take my time here and ask some questions:

1. In the subsection "Searchable History":

> The command history is stored in the file ~/.local/share/fish/fish_history (or $XDG_DATA_HOME/fish/fish_history if that variable is set) by default. However, you can set the fish_history environment variable to change the name of the history session (resulting in a <session>_history file); both before starting the shell and while the shell is running.

In my installation of fish (version 2.7.1-1246-geca4273f-dirty): XDG_DATA_HOME is not defined. Can you tell me more about this variable? It is just mentioned once in the documentation. Can I leave it out of the documentation if it is not part of fish?

2. In the section "Commandline Editor":

I could not get to work some of the keys:

> Altw prints a short description of the command under the cursor. 

it is not the command under the cursor, but before the cursor for me: if I have `echo hello` and have the cursor at the end it does not find anything, even if I mark `echo` with the cursor. Also what I get looks a little weird to me, as if it is too much:

```fish
~> bind k
bind(ntcl)               - Arrange for X events to invoke Tcl scripts
builtin(1), !(1), %(1), .(1), :(1), @(1), {(1), }(1), alias(1), alloc(1), bg(1), bind(1), bindkey(1), break(1), breaksw(1), builtins(1), case(1), cd(1), chdir(1), command(1), complete(1), continue(1), default(1), dirs(1), do(1), done(1), echo(1), echotc(1), elif(1), else(1), end(1), endif(1), endsw(1), esac(1), eval(1), exec(1), exit(1), export(1), false(1), fc(1), fg(1), filetest(1), fi(1), for(1), foreach(1), getopts(1), glob(1), goto(1), hash(1), hashstat(1), history(1), hup(1), if(1), jobid(1), jobs(1), kill(1), limit(1), local(1), log(1), login(1), logout(1), ls-F(1), nice(1), nohup(1), notify(1), onintr(1), popd(1), printenv(1), pushd(1), pwd(1), read(1), readonly(1), rehash(1), repeat(1), return(1), sched(1), set(1), setenv(1), settc(1), setty(1), setvar(1), shift(1), source(1), stop(1), suspend(1), switch(1), telltc(1), test(1), then(1), time(1), times(1), trap(1), true(1), type(1), ulimit(1), umask(1), unalias(1), uncomplete(1), unhash(1), unlimit(1), unset(1), unsetenv(1), until(1), wait(1), where(1), which(1), while(1) - shell built-in commands
dnsextd(8)               - BIND Extension Daemon " Name Description for whatis database
slapo-pbind(5)           - proxy bind overlay to slapd
term_bind(n), term::receive::bind(n) - Keyboard dispatch from terminals
```

> ControlW moves the previous path component (everything up to the previous "/") to the 

I kind of struggle to get what this key does with the killring, I think it should add the last word to the killring, but it sometimes adds more then that, see also [https://github.com/fish-shell/fish-shell/issues/772](https://github.com/fish-shell/fish-shell/issues/772 ). Can anybody help me with that: what exactly is ControlW doing?


3. Missing: a chapter about navigation in the file system:
- using `..` or `~`
- influencing how directories are abriviated
- using the stack of directories with `dirh` and similar commands


I am saving this, so it doesn't get lost, but will continue writing my questions here, okay?

**Comment by faho** [comment]
>XDG_DATA_HOME is not defined. Can you tell me more about this variable? 

It's from the [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html), a sort of "standard" for file locations. It's a bit of a linux-ism, but not wholly unknown elsewhere.

>It is just mentioned once in the documentation

Are you sure about that? It seems to be in the history builtin docs as well. Also, XDG_CONFIG_HOME is mentioned a bunch.

>Can I leave it out of the documentation if it is not part of fish?

I'd rather not leave it out _entirely_. It might be okay to give the simple path here and explain the full thing in a footnote or something, but this is actual behavior that should be documented.

> it is not the command under the cursor, but before the cursor for me

For me, if I do `cd git`, place the cursor somewhere on the "git" (even if it appears like `▮git` or `git▮` in my terminal), and press alt-w, it shows information related to "git". That seems reasonable, and is what I would interpret "under the cursor" as. It's the word the cursor is _on_.

Anyway, this seems like it could do with some changes - it should probably use `__fish_describe_command` (what we use to describe commands for the completions), and it uses a bare `test $tok[1]`, which is a big no-no.

>ControlW moves the previous path component (everything up to the previous "/") to the

>I kind of struggle to get what this key does with the killring,

To me it looks like calling it successively _appends_ to the last killring entry, instead of making a new one. So two ctrl-w in a row on e.g. `a/b/c` (so that the buffer is `a/`), and the top of the killring contains "b/c", instead of just "b/". Not sure if that's supposed to work that way?

Note that the killring isn't used much anymore. #772 talked about it syncing with the x clipboard - which is something it used to do, but which we've removed in favor of the `fish_clipboard_*` functions that you can explicitly invoke.

> Missing: a chapter about navigation in the file system:

Ooh, that sounds nice!

>using .. or ~

This is actually just a special case of "implicit cd". When you enter a directory name instead of a command, we `cd` automatically. Since `..` is a directory, and `~` expands to your home directory, those trigger it as well.

>influencing how directories are abriviated

You mean prompt_pwd?

>I am saving this, so it doesn't get lost, but will continue writing my questions here, okay?

Sure!

**Comment by olets** [comment]
Happened on this because it's `good first issue`. Looks like #5132 more or less addressed it, with additional polish in #5640. Should it be closed?

**Comment by faho** [comment]
Yeah.

This is an ongoing concern (documentation is never truly done), but we've done some good work in the area, so we can close it.

**Comment by faho** [comment]
That means if you find something in the docs that you think should be reworded, especially if it's to make it more understandable or "human" (we've had a bit of an issue with robot voice in our docs), go for it.
